blpapi.ts

```diff
@@ -0,0 +1,728 @@
+
+    // DATA
+    data: any;
+    name: string;
+    message: string;
+    constructor(data: any) {
+        this.data = data;
+        this.name = BlpApiError.NAME;
+        // Subscription errors have a description, other errors have a message.
+        this.message = data.reason.message || data.reason.description;
+    }
+}
+
+// CONSTANTS
+var EVENT_TYPE = {
+    RESPONSE:         'RESPONSE',
+    PARTIAL_RESPONSE: 'PARTIAL_RESPONSE'
+};
+
+// Mapping of request types to response names to listen for.
+// The strings are taken from section A of the BLPAPI Developer's Guide, and are organized by
+// service.
+var REQUEST_TO_RESPONSE_MAP: { [index: string]: string; } = {
+    // //blp/refdata
+    'HistoricalDataRequest': 'HistoricalDataResponse',
+    'IntradayTickRequest':   'IntradayTickResponse',
+    'IntradayBarRequest':    'IntradayBarResponse',
+    'ReferenceDataRequest':  'ReferenceDataResponse',
+    'PortfolioDataRequest':  'PortfolioDataResponse',
+    'BeqsRequest':           'BeqsResponse',
+
+    // //blp/apiflds
+    'FieldInfoRequest':              'fieldResponse',
+    'FieldSearchRequest':            'fieldResponse',
+    'CategorizedFieldSearchRequest': 'categorizedFieldResponse',
+
+    // //blp/instruments
+    'instrumentListRequest': 'InstrumentListResponse',
+    'curveListRequest':      'CurveListResponse',
+    'govtListRequest':       'GovtListResponse',
+
+    // //blp/tasvc
+    'studyRequest':          'studyResponse',
+
+    // //blp/apiauth
+    'TokenGenerationRequest': 'TokenGenerationSuccess',
+    'AuthorizationRequest':   'AuthorizationSuccess'
+};
+
+// Mapping of service URIs to the event names to listen to when subscribed to these services.
+var SERVICE_TO_SUBSCRIPTION_EVENTS_MAP: { [uri: string]: string[]; } = {
+    '//blp/mktdata':  ['MarketDataEvents'],
+    '//blp/mktvwap':  ['MarketDataEvents'],
+    '//blp/mktbar':   ['MarketBarStart', 'MarketBarUpdate', 'MarketBarEnd'],
+    '//blp/pagedata': ['PageUpdate']
+};
+
+// Convert generic BLPAPI errors caused by thrown C++ exceptions to error objects with
+// a type that has the same name than the exception type.
+var getError = (() => {
+    var errorTypeNames = [
+        'DuplicateCorrelationIdException',
+        'InvalidStateException',
+        'InvalidAggrgumentException',
+        'InvalidConversionException',
+        'IndexOutOfRangeException',
+        'FieldNotFoundException',
+        'NotFoundException',
+        'UnknownErrorException',
+        'UnsupportedOperationException'
+    ];
+
+    errorTypeNames.forEach((typeName: string): void => {
+        exports[typeName] = createError(typeName, Error);
+    });
+    return (error: any): Error => {
+        var typeName = error.typeName;
+        if (typeName in exports) {
+            return new exports[typeName](error.message);
+        } else {
+            return error;
+        }
+    };
+})();
+
+function isObjectEmpty(obj: Object): boolean {
```

Comment:
Sorry didn't realize it's old code. It's ok to keep it and we can improve it later on.